lib/icingadb/icingadb-objects.cpp

```diff
@@ -1338,6 +1339,12 @@ bool IcingaDB::PrepareObject(const ConfigObject::Ptr& object, Dictionary::Ptr& a
 			attributes->Set("end_time", TimestampToMilliseconds(downtime->GetFixed() ? downtime->GetEndTime() : (downtime->GetTriggerTime() + downtime->GetDuration())));
 		}
 
+		auto duration = downtime->GetDuration();
+		if (downtime->GetFixed() || ! downtime->IsInEffect()) {
+			duration = downtime->GetEndTime() - downtime->GetStartTime();
+		}
+		attributes->Set("duration", TimestampToMilliseconds(duration));
```

Comment:
Can you please explain why this is set this way, in particular why a flexible downtime should use the scheduled duration when not in effect?

Reply:
I discussed this with Yonas yesterday. We have decided on the scheduled duration here, as the flexible downtime actually ends after this duration if it is not / never in effect. This duration will of course need to be updated when it comes into effect. I don't know if that's part of this PR (as it should).

Reply:
Can we go back a step? #9049 says that you want to sort by this column. Is this just for when a user decides to sort downtimes by duration, or is this needed for some specific view? Maybe it helps to know what this will be used for in the end.

Anyways, so if you have a flexible 1h downtime in a 3h window, you want to day that the downtime has a duration of 1h while it's active, but 3h before and after that? This sounds strange to me.

Reply:
This duration refers to the actual downtime duration, i.e for fixed downtimes `duration` is always `scheduled_end - scheduled_start`, but for flexible downtime it is slightly different. If a flexible downtime is not in effect, then the actual duration is still `scheduled_end - scheduled_start`, but if the downtime is triggered, then the duration is the same as `flexible_duration` which you specified when creating the downtime. And yes this column is used for sorting downtimes by duration and for creating the downtime timeline progress.

<img width="603" alt="Bildschirmfoto 2021-11-05 um 15 54 48" src="https://user-images.githubusercontent.com/57616252/140530751-fa4f85d5-f2ab-4654-87d2-e63d95d44f0e.png">

Reply:
But why should the duration be reset to `scheduled_end - scheduled_start` after a flexible downtime has ended?

Reply:
This is, obviously, wrong.

Reply:
I still don't get why it would be a problem if the duration was always the `flexible_duration` for flexible downtimes. I'd still say it's a 1 hour downtime in a 3 hour window all the time.

However, even if you'd want to go for `scheduled_duration` before it came into effect and `flexible_duration` afterwards, I don't think you would want to check `IsInEffect()` here as this tells you if the current time is within the effective period. I think you might want to check if the downtime has triggered at all. Check for that would be `downtime->GetTriggerTime() != 0`.